linkerd/app/outbound/src/http/handle_proxy_error.rs

```diff
@@ -0,0 +1,195 @@
+use http::{Request, Response};
+use linkerd_app_core::{
+    svc::{layer, NewService, Service},
+    Error,
+};
+use linkerd_proxy_http::ClientHandle;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+#[derive(Clone, Debug)]
+pub struct NewHandleProxyError<N> {
+    inner: N,
+}
+
+#[derive(Clone, Debug)]
+pub struct HandleProxyError<N> {
+    inner: N,
+}
+
+impl<N> NewHandleProxyError<N> {
+    fn new(inner: N) -> Self {
+        Self { inner }
+    }
+
+    pub fn layer() -> impl layer::Layer<N, Service = Self> + Clone + Copy {
+        layer::mk(Self::new)
+    }
+}
+
+impl<T, N> NewService<T> for NewHandleProxyError<N>
+where
+    N: NewService<T>,
+{
+    type Service = HandleProxyError<N::Service>;
+
+    fn new_service(&mut self, target: T) -> Self::Service {
+        let inner = self.inner.new_service(target);
+        HandleProxyError { inner }
+    }
+}
+
+impl<S, A, B> Service<Request<A>> for HandleProxyError<S>
+where
+    S: Service<Request<A>, Response = Response<B>>,
+    S::Response: Send,
+    S::Error: Into<Error> + Send,
+    S::Future: Send + 'static,
+    A: Send + 'static,
+    B: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = Pin<Box<dyn Future<Output = Result<S::Response, Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(Into::into)
+    }
+
+    fn call(&mut self, req: Request<A>) -> Self::Future {
+        let client = req.extensions().get::<ClientHandle>().cloned();
+        debug_assert!(client.is_some(), "Missing client handle");
+        let response = self.inner.call(req);
+        Box::pin(async move {
+            match response.await {
+                Ok(rsp) => {
+                    if let Some(message) =
+                        rsp.headers().get(linkerd_app_core::errors::L5D_PROXY_ERROR)
+                    {
+                        tracing::info!(?message, "response contained L5D_PROXY_ERROR header");
+                        // Gracefully teardown the accepted connection.
+                        if let Some(ClientHandle { close, .. }) = client {
+                            tracing::info!("connection closed");
```

Comment:
not sure if we should have _two_ log lines at the info level for this, since those will be disabled by default. i would probably add something like "response contained ... header, closing connection" to the other log line, and move this one to the debug or trace level just to indicate that the connection was successfully closed for debugging purposes (or remove it entirely?).

Reply:
I made these separate because if for some reason the request does not have the client handle, the connection will not be closed. So the first line handles the fact that the `L5D-PROXY-ERROR` is set, and the other is when we actually close the connection.

The request _should_ have the client handle, but if for some reason it doesn't it could be misleading to say the connection is closed.

> not sure if we should have two log lines at the info level for this, since those will be disabled by default

Won't `info` level be enabled by default? I thought these may be important enough to be `info` instead of `debug`, but I'm also fine changing it!

Reply:
Sorry, I meant to say that info is enabled by default.

If we're so concerned about cases where the `ClientHandle` is missing that it's worth having two logs at the `INFO` level for it...why don't we just remove the `debug_assert!` that it's present, and `expect` it when extracting it from the request's `Extensions`? if it's ever not set, that's _definitely_ a bug, and shouldn't ever be possible, so it's fine to just panic in that case. Then, we don't have to worry about logging the presence of the header and the successful closure of the connection separately.

Reply:
That makes sense. The `debug_asssert!` came from the other location where we expect a `ClientHandle`, but since it should always be set here adding an `expect` sounds good.